samplers/aws/xray/internal/client.go

```diff
@@ -0,0 +1,200 @@
+	ServiceType   string            `json:"ServiceType"`
+	ResourceARN   string            `json:"ResourceARN"`
+	Attributes    map[string]string `json:"Attributes"`
+	ServiceName   string            `json:"ServiceName"`
+	Host          string            `json:"Host"`
+	HTTPMethod    string            `json:"HTTPMethod"`
+	URLPath       string            `json:"URLPath"`
+	ReservoirSize int64             `json:"ReservoirSize"`
+	FixedRate     float64           `json:"FixedRate"`
+	Priority      int64             `json:"Priority"`
+	Version       int64             `json:"Version"`
+}
+
+type getSamplingTargetsInput struct {
+	SamplingStatisticsDocuments []*samplingStatisticsDocument
+}
+
+// samplingStatisticsDocument is used to store current state of sampling data.
+type samplingStatisticsDocument struct {
+	// a unique identifier for the service in hexadecimal
+	ClientID *string
+
+	// the name of the sampling rule
+	RuleName *string
+
+	// the number of requests that matched the rule
+	RequestCount *int64
+
+	// the number of requests borrowed
+	BorrowCount *int64
+
+	// the number of requests sampled using the rule
+	SampledCount *int64
+
+	// the current time
+	Timestamp *int64
+}
+
+// getSamplingTargetsOutput is used to store parsed json sampling targets
+type getSamplingTargetsOutput struct {
+	LastRuleModification    *float64                  `json:"LastRuleModification,omitempty"`
+	SamplingTargetDocuments []*samplingTargetDocument `json:"SamplingTargetDocuments,omitempty"`
+	UnprocessedStatistics   []*unprocessedStatistic   `json:"UnprocessedStatistics,omitempty"`
+}
+
+// samplingTargetDocument contains updated targeted information retrieved from X-Ray service
+type samplingTargetDocument struct {
+	// the percentage of matching requests to instrument, after the reservoir is
+	// exhausted
+	FixedRate *float64 `json:"FixedRate,omitempty"`
+
+	// the number of seconds for the service to wait before getting sampling targets
+	// again
+	Interval *int64 `json:"Interval,omitempty"`
+
+	// the number of requests per second that X-Ray allocated this service
+	ReservoirQuota *int64 `json:"ReservoirQuota,omitempty"`
+
+	// when the reservoir quota expires
+	ReservoirQuotaTTL *float64 `json:"ReservoirQuotaTTL,omitempty"`
+
+	// the name of the sampling rule
+	RuleName *string `json:"RuleName,omitempty"`
+}
+
+type unprocessedStatistic struct {
+	ErrorCode *string `json:"ErrorCode,omitempty"`
+	Message   *string `json:"Message,omitempty"`
+	RuleName  *string `json:"RuleName,omitempty"`
+}
+
+type xrayClient struct {
+	// http client for sending sampling requests to the collector
+	httpClient *http.Client
+
+	// resolved URL to call getSamplingRules API
+	samplingRulesURL string
+
+	// resolved URL to call getSamplingTargets API
+	samplingTargetsURL string
+}
+
+// newClient returns an HTTP client with proxy endpoint.
+func newClient(addr string) (client *xrayClient, err error) {
+	endpoint := "http://" + addr
```

Comment:
I think we might need `http` (local testing) and `https` both so I was just thinking rather than taking endpoint from user we can just directly take host like `http://localhost:2000` or `https://localhost:2000`. Also, I'm not super sure that `awsxrayproxy` supports HTTPS, @anuraaga any comments on that? We can validate the host in `validateConfig` method. @Aneurysm9 does this sound good to you?